src/exporters/otlp/payload.rs

```diff
@@ -0,0 +1,89 @@
+// SPDX-License-Identifier: Apache-2.0
+
+use crate::exporters::http::metadata_extractor::MetadataExtractor;
+use crate::topology::payload::MessageMetadata;
+use bytes::Bytes;
+use http::Request;
+use http_body_util::Full;
+use hyper::body::{Body, Frame, SizeHint};
+use std::pin::Pin;
+use std::task::{Context, Poll};
+use tower::BoxError;
+
+/// OTLP payload wrapper that carries metadata through the HTTP request pipeline
+#[derive(Clone)]
+pub struct OtlpPayload {
+    // Store the body content to avoid cloning data
+    inner: Option<Full<Bytes>>,
+    pub metadata: Option<Vec<MessageMetadata>>,
+    pub size: usize,
+}
+
+impl OtlpPayload {
+    /// Create a new OtlpPayload from a request with metadata
+    pub fn from_request_with_metadata(
+        request: Request<Full<Bytes>>,
+        metadata: Option<Vec<MessageMetadata>>,
+        size: usize,
+    ) -> Request<OtlpPayload> {
+        let (parts, body) = request.into_parts();
+        let payload = Self {
+            inner: Some(body),
+            metadata,
+            size,
+        };
+        Request::from_parts(parts, payload)
+    }
```

Comment:
So it looks like on these payload types we're wrapping `Request<PayloadWithMetadata>` and then implementing the Body trait to extract the body (byte) frames. Just brainstorming, but I wonder if it could also be reversed, so something like:
```
struct MessagePayload<ReqBody> {
  inner: Request<ReqBody>,
  metadata: Option<Vec<MessageMetadata>>,
}
```
Then on the lower client layers, instead of taking a `Request<ReqBody>` they could take a `AsRef<Request<ReqBody>>` and MessagePayload could implement AsRef<Request<ReqBody>> to return inner. Not sure entirely if that could work or if it'd have other hang ups.